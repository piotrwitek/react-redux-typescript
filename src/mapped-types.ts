// tslint:disable:max-line-length
/**
 * Credits to all the people who given inspiration and shared some very usefull code snippets
 * in the following github issue: https://github.com/Microsoft/TypeScript/issues/12215
 */

/**
 * SetDifference
 * @desc Set difference of given literal union types `A` and `B`
 */
export type SetDifference<A extends string, B extends string> = (
  {[P in A]: P } & {[P in B]: never } & { [k: string]: never }
)[A];

/**
 * SetComplement
 * @desc Set complement of given literal union types `A` and it's subset `A2`
 */
export type SetComplement<A extends string, A2 extends A> =
  SetDifference<A, A2>;

/**
 * SymmetricDifference
 * @desc Set difference of the union and the intersection of given literal union types `A` and `B`
 */
export type SymmetricDifference<A extends string, B extends string> =
  SetDifference<A | B, A & B>;

/**
 * Omit
 * @desc From object type `T` remove a set of properties `K`
 */
export type Omit<T extends object, K extends keyof T> = (
  Pick<T, SetComplement<keyof T, K>>
);

/**
 * Diff
 * @desc From object type `T` pick a set of properties that doesn't exist in `U`
 */
export type Diff<T extends object, U extends object> = (
  Pick<T, SetDifference<keyof T, keyof U>>
);

/**
 * Subtract
 * @desc From object type `T` pick a set of properties that doesn't exist in `U`, when `U` is a subtype of `T`
 */
export type Subtract<T extends U, U extends object> = (
  Pick<T, SetComplement<keyof T, keyof U>>
);

/**
 * Overwrite
 * @desc From object type `T` pick a set of properties that doesn't exist in `U` and from `U` pick a set of properties that exist in both `T` & `U`
 */
export type Overwrite<T extends object, U extends object> = (
  Pick<
  (Diff<T, U> & U),
  SetComplement<keyof T, never>
  >
);

/**
 * Assign
 * @desc From object type `T` pick a set of properties that doesn't exist in `U` and from `U` pick all properties
 */
export type Assign<T extends object, U extends object> = (
  Pick<
  (Diff<T, U> & U),
  SetComplement<keyof (T & U), never>
  >
);
