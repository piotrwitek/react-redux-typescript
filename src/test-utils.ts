export function testType<T>(a: T): T { return undefined as any; }

// export interface IsType<T> {
//   (v: T): '1';
//   (v: any): '0';
// }
