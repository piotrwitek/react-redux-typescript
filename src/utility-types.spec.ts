import { testType } from '../utils/test-utils';
import {
  $Call,
  $Keys,
  $Values,
  $ReadOnly,
  $Diff,
  $PropertyType,
  $ElementType,
  $Shape,
  $NonMaybeType,
  Class,
} from './utility-types';
import { _DeepReadonlyObject } from './mapped-types';
/**
 * Fixtures
 */

type Props = { name: string; age: number; visible: boolean };
type DefaultProps = { age: number };

class Foo {}

/**
 * Tests
 */

// @dts-jest:group $Keys
it('$Keys', () => {
  // @dts-jest:pass:snap
  testType<$Keys<Props>>();
});

// @dts-jest:group $Values
it('$Values', () => {
  // @dts-jest:pass:snap
  testType<$Values<Props>>();
});

// @dts-jest:group $ReadOnly
it('$ReadOnly', () => {
  // @dts-jest:pass:snap
  testType<$ReadOnly<Props>>();
});

// @dts-jest:group $Diff
it('$Diff', () => {
  // @dts-jest:pass:snap
  testType<$Diff<Props, DefaultProps>>();
});

// @dts-jest:group $PropertyType
it('$PropertyType', () => {
  // @dts-jest:pass:snap
  testType<$PropertyType<Props, 'name'>>();

  // @dts-jest:pass:snap
  testType<$PropertyType<[boolean, number], '0'>>();
  // @dts-jest:pass:snap
  testType<$PropertyType<[boolean, number], '1'>>();
});

// @dts-jest:group $ElementType
it('$ElementType', () => {
  // @dts-jest:pass:snap
  testType<$ElementType<Props, 'name'>>();

  // @dts-jest:pass:snap
  testType<$ElementType<[boolean, number], 0>>();
  // @dts-jest:pass:snap
  testType<$ElementType<[boolean, number], 1>>();

  // @dts-jest:pass:snap
  testType<$ElementType<boolean[], number>>();

  // @dts-jest:pass:snap
  testType<$ElementType<{ [key: string]: number }, string>>();
});

// @dts-jest:group $Call
it('$Call', () => {
  // @dts-jest:pass:snap
  testType<$Call<(amount: number) => { type: 'ADD'; payload: number }>>();

  type ExtractPropType<T extends { prop: any }> = (arg: T) => T['prop'];
  type Obj = { prop: number };
  type PropType = $Call<ExtractPropType<Obj>>;
  // @dts-jest:pass:snap
  testType<PropType>();
  // type Nope = $Call<ExtractPropType<{ nope: number }>>; // Error: argument doesn't match `Obj`.

  type ExtractReturnType<T extends () => any> = (arg: T) => ReturnType<T>;
  type Fn = () => number;
  type FnReturnType = $Call<ExtractReturnType<Fn>>;
  // @dts-jest:pass:snap
  testType<FnReturnType>();
});

// @dts-jest:group $Shape
it('$Shape', () => {
  // @dts-jest:pass:snap
  testType<$Shape<Props>>();
});

// @dts-jest:group $NonMaybeType
it('$NonMaybeType', () => {
  // @dts-jest:pass:snap
  testType<$NonMaybeType<string|null|undefined>>();
});

// @dts-jest:group Class
it('Class', () => {
  // @dts-jest:pass:snap
  testType<Class<Foo>>();
});
