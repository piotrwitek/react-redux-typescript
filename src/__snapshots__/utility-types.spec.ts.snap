// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$Call testType<$Call<(amount: number) => { type: 'ADD'; payload: number }>>() 1`] = `"{ type: \\"ADD\\"; payload: number; }"`;

exports[`$Call testType<FnReturnType>() 1`] = `"number"`;

exports[`$Call testType<PropType>() 1`] = `"number"`;

exports[`$Diff testType<$Diff<Props, DefaultProps>>() 1`] = `"Pick<Props, \\"name\\" | \\"visible\\">"`;

exports[`$ElementType testType<$ElementType<[boolean, number], 0>>() 1`] = `"boolean"`;

exports[`$ElementType testType<$ElementType<[boolean, number], 1>>() 1`] = `"number"`;

exports[`$ElementType testType<$ElementType<{ [key: string]: number }, string>>() 1`] = `"number"`;

exports[`$ElementType testType<$ElementType<Props, 'name'>>() 1`] = `"string"`;

exports[`$ElementType testType<$ElementType<boolean[], number>>() 1`] = `"boolean"`;

exports[`$Keys testType<$Keys<Props>>() 1`] = `"\\"name\\" | \\"age\\" | \\"visible\\""`;

exports[`$NonMaybeType testType<$NonMaybeType<string|null|undefined>>() 1`] = `"string"`;

exports[`$PropertyType testType<$PropertyType<[boolean, number], '0'>>() 1`] = `"boolean"`;

exports[`$PropertyType testType<$PropertyType<[boolean, number], '1'>>() 1`] = `"number"`;

exports[`$PropertyType testType<$PropertyType<Props, 'name'>>() 1`] = `"string"`;

exports[`$ReadOnly testType<$ReadOnly<Props>>() 1`] = `"_DeepReadonlyObject<Props>"`;

exports[`$Shape testType<$Shape<Props>>() 1`] = `"Partial<Props>"`;

exports[`$Values testType<$Values<Props>>() 1`] = `"string | number | boolean"`;

exports[`Class testType<Class<Foo>>() 1`] = `"Class<Foo>"`;
